Add curve analysis unit support for T1/T2 migration

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

T1 and T2 class have own unit handling logic which is not supported by the curve analysis base class.
To migrate these analysis, this functionality should be supported. This also benefits another analysis subclasses.

### Details and comments

This PR contains two new functionalities for unit handling.

- x value unit support:

If xvalue is provided with circuit metadata `unit`, the prefix is applied. For example, given `metadata = {"xval": 1, "unit": "kHz"}`, x value will be `1e3`.

- result parameter auto scaling

If `result_parameters` option is specified by `ParameterRepr` instance with valid `unit`, the parameter value is automatically scaled with auxiliary unit. For example, given `result_parameters=[ParameterRepr("freq", unit="Hz")]` is set to analysis options and fit parameter value is 1234.5, the `FitVal` object is created with `value=1.2345` and `unit="kHz"`.

yaelbh commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -612,12 +616,23 @@ def _is_target_series(datum, **filters):
+                # apply aux unit (1 ms -> 0.001 s)
+                xval = apply_prefix(xval, datum["metadata"].get("unit", None))
+            except Exception:
+                pass
```
We want to catch invalid units

yaelbh commented on releasenotes/notes/add-aux-unit-support-curve-analysis-beb4150ef5de283b.yaml:
```diff
@@ -0,0 +1,6 @@
+  - |
+    Parameter reported by :py:class`qiskit_experiments.curve_analysis.curve_analysis.CurveAnalysis`
+    subclass can be scaled with auxiliary unit if unit is provided. For example,
+    Frequency = 1000 ± 0.1 Hz is now displayed as 1 ± 0.0001 kHz.
```
If `unit` is provided and `scale` is `True`. By default `scale` is False, which is the desired behavior. In fact, where do you allow the user to control `scale`?

nkanazawa1989 replied:
Now scale is removed from fitval method.

yaelbh commented on test/curve_analysis/test_curve_fit.py:
```diff
@@ -63,6 +64,67 @@ class TestAnalysis(CurveAnalysis):
+        self.assertEqual(a_val, FitVal(2.0, 0.2))
+
+        a_val = data.fitval("c")
+        self.assertEqual(a_val, FitVal(3.0, 0.3))
```
I guess that you wanted to call the variables `b_val` and `c_val`

nkanazawa1989 replied:
Good catch!

yaelbh commented on test/curve_analysis/test_curve_fit.py:
```diff
@@ -63,6 +64,67 @@ class TestAnalysis(CurveAnalysis):
+            FitVal(10.0, unit="MHz"),
+        ]
+
+        for i in range(15):
```
15 is hard-coded, replace with `len(unitrefs)` or `len(popt)` (maybe verify that they're equal?)

nkanazawa1989 replied:
This test is removed since scaling is not performed with `FitData`

yaelbh reviewed (changes requested):
Thanks @nkanazawa1989 for taking care of it. When the time comes to make T1 ad T2* inherit from `CurveAnalysis`, please coordinate with me and @merav-aharoni, because I also started to work on it.

1. In the release notes, as well as in the description of the PR, you state that scaling of the result is automatically performed if unit is provided. However you set `scale` to `False`. Note that we normally want `scale` to be `False`, because we have an informal spec (derived from discussions mainly with Oliver) to display the results in SI (except for the figure, where we are encouraged to scale).
2. Need to handle `dt`.
3. Currently T1 result contains a field named `circuit_unit`, which disappears in this PR, but I'm not sure if we really need this field.

nkanazawa1989 commented:
Thanks @yaelbh , I wasn't going to takeover what you're currently doing (happy to help! 😄). I carefully read the T1 and T2 implementation and realized that unit handling seems to be a blocker.

> Note that we normally want scale to be False, because we have an informal spec (derived from discussions mainly with Oliver) to display the results in SI (except for the figure, where we are encouraged to scale).

This is what I really misunderstood. I realized you only used the unit conversion for visualization, but I applied it for returned parameters. It makes me think we just need to move unit conversion functionality to terra and just need to call it from visualization method. See this PR https://github.com/Qiskit/qiskit-terra/pull/6885.

I'll fix the implementation accordingly.

nkanazawa1989 commented:
@yaelbh The PR is updated. Main change is
- Remove added `scale` option from `FitData.fitval` method. Return parameters should be no prefix.
- Add `xval_unit` `yval_unit` support in analysis options. This will be applied to visualization.

> Need to handle dt.

I think this should be managed by circuit method of experiment. The circuit metadata is only used by the post-analysis, thus formatting metadata itself is not quite strange. In the practical reason, these data extraction will be eventually delegated to `DataProcessor` and it might be difficult to write custom data formatting routine only for T1 or T2 experiment (especially this conversion requires backend object, or `dt`, and needs API change). If metadata itself is already formatted, the post processing logic is drastically simplified. 

> Currently T1 result contains a field named circuit_unit, which disappears in this PR, but I'm not sure if we really need this field.

I added this to analysis options as `xval_unit` rather than analysis result metadata. Because we don't keep raw x-values in any database entry, the `circuit_unit` is no longer convenient to end-users. Note that `T1` entry is now `FitVal` object that already contains unit in it.

yaelbh reviewed (commented):
I'm fine with handling `dt` at the experiment side, like you do here. I think the original delay times (in `dt`), and/or the backend's `dt` parameter, should appear in the analysis result. We can ask experimentalists about it.

We need to coordinate with the IQX team. Currently they take the value, variance, and unit, and compose a representation, just like you do in the code here. If variance can take a unit different from the value's, then they have to know about it and support it. In my opinion the best thing to do is that they stop looking at the value-variance-unit, and just display the string that you provide.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -318,6 +352,10 @@ def _default_options(cls) -> Options:
+            xval_unit (str): Unit of x values. No auxiliary unit is needed here.
+                The value is automatically scaled if unit is provided.
+            yval_unit (str): Unit of y values. No auxiliary unit is needed here.
+                The value is automatically scaled if unit is provided.
```
Added more explanation in https://github.com/Qiskit/qiskit-experiments/pull/305/commits/a17e15dbe854dd90d5656cee4d2de67c80960c70

nkanazawa1989 commented on qiskit_experiments/library/characterization/t1.py:
```diff
@@ -97,16 +98,21 @@ def circuits(self, backend: Optional[Backend] = None) -> List[QuantumCircuit]:
-            AttributeError: if unit is dt but dt parameter is missing in the backend configuration
+            AttributeError: If unit is ``dt``, but ``dt`` parameter is missing in
+                the backend configuration.
+            ValueError: If invalid unit is specified, i.e. not ``dt`` or ``s``.
```
This change is reverted

nkanazawa1989 commented on qiskit_experiments/library/characterization/t2ramsey.py:
```diff
@@ -104,17 +104,19 @@ def circuits(self, backend: Optional[Backend] = None) -> List[QuantumCircuit]:
-            AttributeError: if unit is 'dt', but 'dt' parameter is missing in the backend configuration.
+            AttributeError: If unit is ``dt``, but ``dt`` parameter is missing in
+                the backend configuration.
+            ValueError: If invalid unit is specified, i.e. not ``dt`` or ``s``.
```
This change is reverted

nkanazawa1989 commented on releasenotes/notes/add-aux-unit-support-curve-analysis-beb4150ef5de283b.yaml:
```diff
@@ -0,0 +1,12 @@
+    To enable this change, `xval_unit` and `yval_unit` are added to the analysis options of
+    :py:class`qiskit_experiments.curve_analysis.curve_analysis.CurveAnalysis`.
+    These options are likely set by the class itself rather than end-users,
+    and implicitly update the fit plots.
```
I agree each experiment module level change is not significant, but base class change should be described in the reno otherwise community people cannot find how to use new functionality in their module (especially we expect people use experiment to develop their library)

nkanazawa1989 commented:
> I think the original delay times (in dt), and/or the backend's dt parameter, should appear in the analysis result. We can ask experimentalists about it.

I agree with this, however, other modules don't store the raw x values. Actually curve analysis has an option 
https://github.com/Qiskit/qiskit-experiments/blob/2d412864c9609bbd0888dbfc9c58eec26796fad8/qiskit_experiments/curve_analysis/curve_analysis.py#L328
that generates an entry of raw x-y values from post analysis. This is very important to track, for example, TLS error -- we usually draw heatmap of Ramsey experiment at different time. However, we decided to disable this option by default, after some discussion with @chriseclectic .

I added units to extra field of raw values data entry, as you recommended.
https://github.com/Qiskit/qiskit-experiments/pull/305/commits/08dd82d308d6264d994939c1e912d7fae23b143d

> I'm fine with handling dt at the experiment side

After careful investigation of this, I found another issue.

```
delays_dt = np.linspace(1, 100, 1, dtype=int)
exp = T1(0, delays, unit="dt")
exp.analysis_options(t1_guess=30)
```

In this example, we used to specify the `t1_guess` in units of one in the constructor. However, if we implicitly calculate prefactor in the circuit method, we induce inconsistency of unit between initial guess and xvalues in the circuit metadata.
I think the problem is bit complicated and we need to think some logic that passes prefactor to analysis as an analysis options, rather than the circuit metadata (i.e. circuit metadata can vary, but prefactor applied to guess should be defined at experiment level). Perhaps this will be separate PR. Due to this consideration, I reverted the change in T1 and T2 to merge CurveAnalysis side first.

yaelbh reviewed (approved):
Thanks @nkanazawa1989 :smile:
